src/hotspot/share/utilities/vmError.hpp

```diff
@@ -213,4 +213,27 @@ class VMError : public AllStatic {
   static int prepare_log_file(const char* pattern, const char* default_pattern, bool overwrite_existing, char* buf, size_t buflen);
 
 };
+
+class VMErrorCallback {
+  friend class VMError;
+  friend class VMErrorCallbackMark;
+
+  // Link through all callbacks active on a thread
+  VMErrorCallback* _next;
+
+  // Called by VMError reporting
+  virtual void call(outputStream* st) = 0;
+
+public:
+  VMErrorCallback() : _next(nullptr) {}
+};
+
+class VMErrorCallbackMark : public StackObj {
+  Thread* _thread;
```

Comment:
Why would we need the thread here? Why not use Thread::current in dtor? This object is only used as stack object, right?

Reply:
I was treading in Runtime code and Coleen usually wants to use cached-away Thread pointers instead of calling Thread::current() repeatedly. I'm fine with either solution.

Reply:
Given the context it would have to be `Thread::current_or_null_safe()`. But yes we prefer not to re-materialize the current thread if we already have it at hand.

Reply:
Could you explain why it would have to be `Thread::current_or_null_safe()`? The constructor and destructor are run in "normal" JVM code and not in the error handler.

Reply:
Sorry my mistake.